### PR TITLE
Use pre steps to determine whether to record live tests

### PR DESCRIPTION
**Pull request by benbp**
Previously we used template expressions, which are rendered at compile time, in order to set the condition expression (https://github.com/azure/azure-sdk-for-net/blob/c6f9946dbb30ccbf11460e14655175cae50e0fc4/eng/pipelines/templates/jobs/archetype-sdk-tests-jobs.yml#L39). With recent changes, in order to support MaxParallel, we have switched to using a matrix strategy. This change meant that platforms were enumerated via the matrix definition, instead of an `${{ each platforms ...}}` loop in the job definition. As a result, the TestMode setting was then available as a variable instead of a parameter, like so:

```
- job:
  strategy:
    matrix: ${{ parameters.Platforms }}
```
which becomes:
```
- job:
  strategy:
    matrix:
      Windows_NetCore_Record:
        TestMode: Record
```

The pipelines evaluation logic does not have the matrix variables within a template in scope at the time it evaluates a job condition. I think this is because it is in the execution context for job orchestration, which is prior to matrix expansion. However, it DOES have them in scope at the time it evaluates a step condition (since it is the runtime execution context).

See https://developercommunity.visualstudio.com/content/problem/1145638/azure-devops-pipelines-variable-not-expanded-in-te.html and https://docs.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml#use-a-template-parameter-as-part-of-a-condition for some detail on this behavior.

Due to these constraints, I've moved the Record condition into the step condition, and instead of having a separate Windows_NetCore_Record platform, made it so that the Windows_NetCore platform will record and publish recordings when the Record variable is set for the pipeline run. A drawback to this change is that when running a pipeline with Record set to true, it will run the tests for all 6 platforms, instead of just the recording platform (however it will still only record for the intended platform). I don't think this is tradeoff is too bad since the recording will take the same amount of time as before, we'll just consume a few more agents from the queue for the test duration.

Example passing pipeline with Record set to true: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=625512&view=results

### CODE REVIEW

**Comment by benbp** [comment]
Actually there's still a shortcoming with this PR, due to this setting on the Build and Test step:

```
AZURE_TEST_MODE: "${{ coalesce(variables['TestMode'], 'None') }}"
```

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml [hunk 1]
```diff
@@ -86,7 +82,7 @@ parameters:
   type: object
   default:
     - task: CopyFiles@2
-      condition: and(succeededOrFailed(), eq(variables['Agent.JobName'], 'Windows_NetCore_Record'))
+      condition: and(succeededOrFailed(), eq(variables['Record'], 'true'), eq(variables['Agent.JobName'], 'Windows_NetCore'))
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-tests.yml:85]:
```suggestion
      condition: and(succeededOrFailed(), eq(variables['TestMode'], 'RECORD'))
```

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml [hunk 2]
```diff
@@ -95,7 +91,7 @@ parameters:
         contents: 'sdk/$(ServiceDirectory)/**/SessionRecords/**/*.json'
         targetFolder: '$(Build.ArtifactStagingDirectory)/SessionRecords'
     - task: PublishBuildArtifacts@1
-      condition: and(succeededOrFailed(), eq(variables['Agent.JobName'], 'Windows_NetCore_Record'))
+      condition: and(succeededOrFailed(), eq(variables['Record'], 'true'), eq(variables['Agent.JobName'], 'Windows_NetCore'))
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-sdk-tests.yml:94]:
```suggestion
       condition: and(succeededOrFailed(), eq(variables['TestMode'], 'RECORD'))
```

**Review by weshaggard — APPROVED** [review]
A suggestion for the conditions but I think this will work until we can get our other work with the matrix filtering figured out.

**Comment by pakrym** [comment]
> A drawback to this change is that when running a pipeline with Record set to true, it will run the tests for all 6 platforms, instead of just the recording platform (however it will still only record for the intended platform). I don't think this is tradeoff is too bad since the recording will take the same amount of time as before, we'll just consume a few more agents from the queue for the test duration.

Yeah, this is a bit unfortunate and something we tried to avoid with the first iteration. We would need to update the collection script so it doesn't wait for all builds inside a run before downloading the artifacts.

**Pull request merged**